exist-core/src/main/java/org/exist/xquery/value/Type.java

```diff
@@ -105,95 +105,136 @@
     public final static int MAP = 102;
     public final static int ARRAY = 103;
     private final static Logger LOG = LogManager.getLogger(Type.class);
+
+    private static int NO_SUCH_VALUE = -99;
+
     private final static int[] superTypes = new int[512];
-    private final static Int2ObjectMap<String[]> typeNames = new Int2ObjectOpenHashMap<>(100);
-    private final static Object2IntMap<String> typeCodes = new Object2IntOpenHashMap<>(100);
+    private final static Int2ObjectOpenHashMap<String[]> typeNames = new Int2ObjectOpenHashMap<>(64, Hash.FAST_LOAD_FACTOR);
+    private final static Object2IntOpenHashMap<String> typeCodes = new Object2IntOpenHashMap<>(100, Hash.FAST_LOAD_FACTOR);
     static {
-        typeCodes.defaultReturnValue(-1);
+        typeCodes.defaultReturnValue(NO_SUCH_VALUE);
+    }
+    private final static Int2ObjectMap<IntArraySet> unionTypes = new Int2ObjectArrayMap<>(1);
+    private final static Int2IntOpenHashMap primitiveTypes = new Int2IntOpenHashMap(44, Hash.FAST_LOAD_FACTOR);
+    static {
+        primitiveTypes.defaultReturnValue(NO_SUCH_VALUE);
     }
 
     static {
+        // ANY types
         defineSubType(ANY_TYPE, ANY_SIMPLE_TYPE);
         defineSubType(ANY_TYPE, UNTYPED);
+
+        // ANY_SIMPLE types
         defineSubType(ANY_SIMPLE_TYPE, ATOMIC);
-        defineSubType(NODE, ELEMENT);
-        defineSubType(NODE, ATTRIBUTE);
-        defineSubType(NODE, TEXT);
-        defineSubType(NODE, PROCESSING_INSTRUCTION);
-        defineSubType(NODE, COMMENT);
-        defineSubType(NODE, DOCUMENT);
-        defineSubType(NODE, NAMESPACE);
-        defineSubType(NODE, CDATA_SECTION);
+        defineSubType(ANY_SIMPLE_TYPE, NUMBER);
 
-        //THIS type system is broken - some of the below should be sub-types of ANY_SIMPLE_TYPE
-        //and some should not!
+        // ITEM sub-types
         defineSubType(ITEM, ATOMIC);
-        defineSubType(ATOMIC, STRING);
-        defineSubType(ATOMIC, BOOLEAN);
-        defineSubType(ATOMIC, QNAME);
+        defineSubType(ITEM, FUNCTION_REFERENCE);
+        //defineSubType(ITEM, NODE);                // TODO(AR) this appears in the XDM 3.1, but uncommenting this breaks a lot of stuff in eXist-db
```

Comment:
Setting NODE to a different value (I tried 7, the first available number not yet taken) allows this definition.
https://github.com/eXist-db/exist/pull/3363/files#diff-14a14d3b3adcc1ebc87fe76d1e77c862R43
At least all Xquery3 Tests pass.

Reply:
Hmmm... okay that's interesting. I would have thought that changing its value should have no effect! Thanks Juri, I will take a look at this...

Reply:
It was the only negative value, therefore I gave it try.

Reply:
If this is turns out to be feasible I would suggest to change NO_VALUE back to -1.

Reply:
Sure. It should not matter if it is positive or negative as long as it is a valid java int.

Reply:
Well with it being negative nothing works... the test suites cannot even be started. With it being zero an exception is thrown that zero is not allowed. When set to a positive value the DB runs and can be used normally and core tests pass.

Reply:
@line-o Okay I checked out what you are mentioned.

I can indeed see the issue that you mention, it is because the Type is used in some places as an index into an array, and that can't be a negative value.

Unfortunately we cannot fix such a thing without changing the format of some `.dbx` files, as the value of the Type is encoded into key/values that are stored in the on-disk files. I haven't checked them all, but the first and most obvious is the `NativeValueIndex`. 

So if we wanted to fix that it would have to happen in a major release change and we would need to bump the file-version numbers of the relevant on-disk file formats.

I think if we get this one merged, we can open a second PR to do that for a 6.0.0 release or whatever